plenum/server/replica.py

```diff
@@ -1451,14 +1467,29 @@ def stashCheckpoint(self, ck: Checkpoint, sender: str):
         self.stashedRecvdCheckpoints[seqNoStart, seqNoEnd][sender] = ck
 
     def processStashedCheckpoints(self, key):
-        i = 0
-        if key in self.stashedRecvdCheckpoints:
-            for sender, ck in self.stashedRecvdCheckpoints[key].items():
-                self.processCheckpoint(ck, sender)
-                i += 1
-        logger.debug('{} processed {} stashed checkpoints for {}'.
-                     format(self, i, key))
-        return i
+
+        if key not in self.stashedRecvdCheckpoints:
+            logger.debug("{} have no stashed checkpoints for {}")
+            return 0
+
+        total_processed = 0
+        senders_of_completed_checkpoints = []
+
+        for sender, checkpoint in self.stashedRecvdCheckpoints[key].items():
+            if self.processCheckpoint(checkpoint, sender):
+                senders_of_completed_checkpoints.append(sender)
+            total_processed += 1
+
+        for sender in senders_of_completed_checkpoints:
+            # unstash checkpoint
+            del self.stashedRecvdCheckpoints[key][sender]
+
```

Comment:
delete `self.stashedRecvdCheckpoints[key]` is `self.stashedRecvdCheckpoints[key]` is empty